src/CustomIndexing/CustomIndexing.jl

```diff
@@ -0,0 +1,130 @@
+    if !haskey(io, :compact)
+        io = IOContext(io, :compact => true)
+    end
+    for (i, (key, value)) = enumerate(x.data)
+        if print_entry(i)
+            print(io, "  ", '[', rpad(key_string(key), pad), "]  =  ", value)
+            if i != length(x)
+                println(io)
+            end
+        elseif i == half_screen_rows
+            println(io, "   ", " "^pad, "   \u22ee")
+        end
+    end
+end
+
+############
+# Indexing #
+############
+
+Base.setindex!(d::SparseArray, value, idx) = setindex!(d.data, value, idx)
+Base.getindex(d::SparseArray, idx) = getindex(d.data, idx)
+Base.eachindex(d::SparseArray) = keys(d.data)
+
+# Need to define it as indices may be non-integers
+Base.to_index(d::SparseArray, idx) = idx
+
+# Arbitrary typed indices. Linear indexing not supported.
+struct IndexAnyCartesian <: Base.IndexStyle end
+Base.IndexStyle(::IndexAnyCartesian, ::IndexAnyCartesian) = IndexAnyCartesian()
+Base.IndexStyle(::Type{<:SparseArray}) = IndexAnyCartesian()
+# eachindex redirect to keys
+Base.keys(::IndexAnyCartesian, d::SparseArray) = keys(d)
+
+################
+# Broadcasting #
+################
+
+# Need to define it as indices may be non-integers
+Base.Broadcast.newindex(d::SparseArray, idx) = idx
+
+struct BroadcastStyle{N} <: Broadcast.BroadcastStyle end
+# TODO throw an helpful error when trying to mix SparseArray with a non-scalar
+# Scalars can be used with SparseArray in broadcast
+Base.BroadcastStyle(::BroadcastStyle{N}, ::Base.Broadcast.DefaultArrayStyle{0}) where {N} = BroadcastStyle{N}()
+Base.BroadcastStyle(::Type{<:SparseArray{T, N}}) where {T, N} = BroadcastStyle{N}()
+function Base.similar(b::Base.Broadcast.Broadcasted{BroadcastStyle{N}}, ::Type{T}) where {T, N, Ax}
+    SparseArray{T, N}(Dict{Tuple, T}())
+end
+
+# Check that all SparseArrays involved have the same indices. The other
+# arguments are scalars
+function check_same_eachindex(each_index) end
+check_same_eachindex(each_index, not_sa, args...) = check_same_eachindex(eachindex, args...)
+function check_same_eachindex(each_index, sa::SparseArray, args...)
+    if Set(each_index) != Set(eachindex(sa))
+        throw(ArgumentError("Cannot broadcast CustomIndexing.SparseArray with different indices"))
+    end
+    check_same_eachindex(eachindex, args...)
+end
+_eachindex(not_sa, args...) = _eachindex(args...)
+function _eachindex(sa::SparseArray, args...)
+    each_index = eachindex(sa)
+    check_same_eachindex(each_index, args...)
+    return each_index
+end
+# Need to define it as it falls back to `axes` by default
+function Base.eachindex(bc::Base.Broadcast.Broadcasted{<:BroadcastStyle})
+    return _eachindex(bc.args...)
+end
+# Need to define it as `axes` is not defined
+Base.Broadcast.instantiate(bc::Base.Broadcast.Broadcasted{<:BroadcastStyle}) = bc
+# Need to define it as indices are not integers
+function Base.copyto!(dest::SparseArray{T, N}, bc::Base.Broadcast.Broadcasted{BroadcastStyle{N}}) where {T, N}
```

Comment:
Is this part of the normal broadcasting interface? Will the code break between minor Julia releases?

Reply:
Yes, `copyto!` and `instantiate` are the place where you are expected to add new methods for new `BroadcastStyle`, it is written in their docstrings